src/gbinder_driver.c

```diff
@@ -447,9 +447,21 @@ gbinder_driver_handle_transaction(
             tx.code, tx.flags, &status);
         break;
     case GBINDER_LOCAL_TRANSACTION_SUPPORTED:
-        reply = gbinder_handler_transact(h, obj, req, tx.code, tx.flags,
-            &status);
-        break;
+        {
+            if(handler)
+            {
+                // this function will be called if we have none null handler
+                reply = gbinder_handler_transact(handler, obj, req, tx.code, tx.flags,
+                    &status);
+            }
+            else
+            {
+                // this function will be called during other transactions processing to allow
+                // local object to handle all callbacks if we have null handler
```

Comment:
This comment should be rewritten. There's no need to explain that we get here with a NULL handle. It's obvious to anyone who can read C code. There's also no need to explain that the function called `gbinder_local_object_handle_transaction` allows local object to handle a transaction.

What needs to be explained is WHY it needs to be called here and WHY it's guaranteed that it's happening on the main thread.

Reply:
Hmm, `gbinder_driver_transact` always calls `gbinder_driver_handle_transaction` with a NULL handler. And it's called from a transaction thread:
```
(gdb) bt
#0  gbinder_driver_transact (self=0x62b2f0, reg=0x62c160, handle=0, code=1, req=0x72ac00, reply=0x0) at src/gbinder_driver.c:970
#1  0x000000000040da92 in gbinder_ipc_tx_internal_exec (priv=0x72b830) at src/gbinder_ipc.c:1298
#2  0x000000000040dd87 in gbinder_ipc_tx_proc (data=0x72b830, object=0x62c1b0) at src/gbinder_ipc.c:1423
#3  0x00007ffff76e294c in g_thread_pool_thread_proxy (data=<optimized out>) at /build/glib2.0-WFyp2b/glib2.0-2.40.2/./glib/gthreadpool.c:307
#4  0x00007ffff76e1fc5 in g_thread_proxy (data=0x72c000) at /build/glib2.0-WFyp2b/glib2.0-2.40.2/./glib/gthread.c:764
#5  0x00007ffff6c4f184 in start_thread (arg=0x7ffff6c46700) at pthread_create.c:312
#6  0x00007ffff73a903d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```
libgbinder must guarantee that all callbacks are invoked on the main thread. Its public API is not thread safe (on purpose, it's not an accident or an oversight).

Reply:
So if we call `gbinder_driver_handle_transaction` with NULL handler actually callback will never be called? 
https://github.com/mer-hybris/libgbinder/blob/master/src/gbinder_handler.h#L60

Reply:
Hmm, glib event loop could also be started not from main thread. Will transactions also be thread safe?

Reply:
Yes, such transactions (valid incoming transactions received while we're waiting for the completion of a transaction that we have submitted) are currently being rejected, it's a bug and it needs to be fixed. Good catch, actually! This IS an oversight.

The context returned by `g_main_context_default()` is assumed to be a set of sources for the event loop running on the main thread. That's the thread where everything is happening (except for what's happening on looper/transaction threads which are being handled veeeery carefully and which don't call any external code and therefore are not visible from outside of libgbinder).

Actually, refcounting API is thread safe but the rest of public API shouldn't be expected to be thread safe (even though parts of it may be).

BTW, this event loop stuff had been recently abstracted away by `GBinderEventLoopIntegration` API (primarily to deal with Qt event loops that are not based on GLib event loop).